agora/Agora/Effect/README.md

```diff
@@ -0,0 +1,163 @@
+  ptryFrom' opq cont =
+    -- this will need to not use punsafeCoerce...
+    cont (punsafeCoerce opq, ())
+```
+
+> All informations above is **well** documented in
+> [Plutonomicon/plutarch](https://github.com/Plutonomicon/plutarch/tree/master/docs)
+> repository. Well, except `PTryFrom`...
+
+## Effect Validator Boilerplate
+
+In Agora, Effects can be built with the given `makeEffect`
+function. This function is a simple boilerplate that checks the GAT
+is handled correctly. This means that the creator of the effect needs
+to only consider the implementation of their effect logic.
+
+``` haskell
+makeEffect ::
+  forall (datum :: PType).
+  (PIsData datum, PTryFrom PData datum) =>
+  CurrencySymbol ->  
+  -- ^ The Currency Symbol of DAO system
+  (forall (s :: S). Term s PCurrencySymbol -> Term s datum -> Term s PTxOutRef -> Term s (PAsData PTxInfo) -> Term s POpaque) ->
+  -- ^ Validator logic
+  ClosedTerm PValidator
+```
+
+An example effect validator would be:
+
+``` haskell
+effectValidator :: forall {s :: S}. CurrencySymbol -> Term s PValidator
+effectValidator currSymbol = makeEffect currSymbol $
+  \_cs (datum' :: Term _ PYourEffectDatum) _txOutRef _txInfo -> P.do
+    opaque (pconstant ())
+```
+
+Again, conveniently, Script Context is already stripped into `PTxOutRef`
+and `PTxInfo`. `PTxInfo` will provide all informations about the
+transaction: inputs, outputs, minted tokens, fees, and more. `PTxOutRef`
+is the output reference for the effect signer, the one who starts the
+effect by burning the GAT. It is easy to overlook it as useless, but, in
+fact, It is very useful for
+
+-   finding the address of effect script
+-   checking who started the effect
+-   accessing other informations
+
+> Utility functions handling `PTxInfo` and `PTxInfo` is provided in
+> Agora/Util.hs
+
+## Effect Validator Logic
+
+The most important piece of the validator has still not been
+discussed: the validator logic. As explained above, validators ensure
+transactions are built correctly and behave as desired.
+
+Normal onchain scripts target to be less strict in order to be useful
+in various cases. That is not the case for effects. Effects need to be
+specific to correctly filter out all ill-formed transactions. Being
+less strict could easily impose vulnerability to entire system.
+
+Whilst the logic of most validators will be specific to those
+validators, there are some general points which should be considered
+in the construction of every effect. Utility functions in `Agora.Util`
+will assist a developer in ensuring that their effect abide by these
+considerations. These include:
+
+-   All effect transactions should burn *exactly one* GAT.
+-   No transactions should result in funds being paid to an effect.
+-   Effects concerning configuration of the governance system should
+    not permit funds to be transferred between outputs.
+-   Needed informations for effect should be explicitly provided by
```

Comment:
information is both plural and singular. We don't say 'informations' in English.